rfd/0033-sni-alpn-teleport-proxy-routing.md

```diff
@@ -0,0 +1,160 @@
+                             └────────────────────────────┘    └─────────────────────────────────┘
+                                            │                                   │                 
+                                            │                                   │                 
+                                            │                                   │                 
+                                            │                                   │                 
+                                            │                                   │                 
+                                            │                                   │                 
+                                            │                                 ALPN:               
+                                            │                         http1.1, h2, default        
+           ┌────────────────┬───────────────┴──┬───────────────┐                │                 
+           │                │                  │               │                │                 
+           │                │                  │               │                │                 
+         ALPN:              │                  │ALPN:          │                │                 
+    teleport-mysql          │            teleport-proxy-ssh    │                │                 
+           │                │                  │               │                │                 
+           │                │ALPN:             │             ALPN:              │                 
+           │           teleport-postgres       │     http1.1, h2, default       │                 
+           │                │                  │               │                │                 
+           ▼                ▼                  ▼               ▼                ▼                 
+    ┌────────────┐  ┌───────────────┐  ┌───────────────┐  ┌─────────┐  ┌────────────────┐         
+    │MySQL proxy │  │Postgres proxy │  │ssh proxy      │  │web proxy│  │kube proxy      │         
+    │service     │  │service        │  │service        │  │service  │  │service         │         
+    └────────────┘  └───────────────┘  └───────────────┘  └─────────┘  └────────────────┘         
+
+
+```
+
+
+### Reducing the number of teleport proxy port by TLS ALPN Routing:
+
+ALPN (Application-Layer Protocol Negotiation) is a TLS extension that allows the application layer to negotiate which protocol should be performed over a TLS connection. Teleport proxy will leverage ALPN routing in order to reduce the number of exposed ports.
+
+Outgoing traffic send to the ssh-proxy, reverse tunnel, MySQL, Postgres and MongoDB proxy services will be wrapped in TLS protocol where the teleport proxy clients (tsh, teleport internal libs, local db proxy) will be responsible for setting one of the following teleport protocols:
+- teleport-ssh-proxy
+- teleport-reverse-tunnel
```

Comment:
Could you add a separate section on how reverse tunnel subsystem will work? It's a little different from others since it's internal teleport service and won't go through local proxy so I feel it's worth expanding on in a dedicated section.

Reply:
added